cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java

```diff
@@ -17,10 +17,170 @@
 
-import com.google.common.util.concurrent.Service;
+import co.cask.cdap.app.runtime.ProgramRuntimeService;
+import co.cask.cdap.app.runtime.ProgramStateWriter;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.internal.app.store.RunRecordMeta;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicate;
+import com.google.common.util.concurrent.AbstractIdleService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
 
 /**
- * A service interface that defines the behavior when run records are corrected
+ * The base implementation of the service that periodically scans for program runs that no longer running.
  */
-public interface RunRecordCorrectorService extends Service {
+public abstract class RunRecordCorrectorService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RunRecordCorrectorService.class);
+  private static final Set<ProgramRunStatus> NOT_STOPPED_STATUSES = EnumSet.of(ProgramRunStatus.STARTING,
+                                                                               ProgramRunStatus.RUNNING,
+                                                                               ProgramRunStatus.SUSPENDED);
+  private final Store store;
+  private final ProgramStateWriter programStateWriter;
+  private final ProgramRuntimeService runtimeService;
+  private final long startTimeoutSecs;
+  private final int txBatchSize;
+
+  RunRecordCorrectorService(CConfiguration cConf, Store store, ProgramStateWriter programStateWriter,
+                            ProgramRuntimeService runtimeService) {
+    this(store, programStateWriter, runtimeService, 2L * cConf.getLong(Constants.AppFabric.PROGRAM_MAX_START_SECONDS),
+         cConf.getInt(Constants.AppFabric.PROGRAM_RUNID_CORRECTOR_TX_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  RunRecordCorrectorService(Store store, ProgramStateWriter programStateWriter, ProgramRuntimeService runtimeService,
+                            long startTimeoutSecs, int txBatchSize) {
+    this.store = store;
+    this.programStateWriter = programStateWriter;
+    this.runtimeService = runtimeService;
+    this.startTimeoutSecs = startTimeoutSecs;
+    this.txBatchSize = txBatchSize;
+  }
+
+  void fixRunRecords() {
+    Set<ProgramRunId> fixed = doFixRunRecords();
+
+    if (!fixed.isEmpty()) {
+      LOG.info("Corrected {} of run records in {} status but no actual program running.",
```

Comment:
Can you improve this message? (The params are reversed and the grammar is odd). Perhaps:

"Correct {n} run records with status in {statuses} that have no actual running program. Such programs likely have crashed or were killed by external signal. "

Reply:
Sure. This was copied without much thought.